### DISCUSSION TITLE
How to enable physical keyboard and physical mouse simulation?

### DISCUSSION THREAD

**Discussion by arbdevml** [NetrisTV/ws-scrcpy#231] [Q&A]
Hello.
First of all very big thank you for this project.
How to enable physical keyboard and physical mouse simulation?

scrcpy supported physical keyboard and physical mouse simulation

Keyboard:
```
scrcpy --hid-keyboard
scrcpy -K  # short version
```
https://github.com/Genymobile/scrcpy#physical-keyboard-simulation-hid

Mouse:
```
scrcpy --hid-mouse
scrcpy -M  # short version
```
https://github.com/Genymobile/scrcpy#physical-mouse-simulation-hid

This feature is very important to control a device using a keyboard and a mouse
because on some devices impossible to enable an additional option
https://github.com/Genymobile/scrcpy/issues/70#issuecomment-373286323

Thank you for your answer.

**Comment by drauggres** [comment]
Hi.

HID support is a part of the scrcpy client (part that runs on pc/mac). In this project I use only the scrcpy server (runs on device).
To support HID in ws-scrcpy someone will need to reimplement [this](https://github.com/Genymobile/scrcpy/pull/2632) for node.js.

> This feature is very important to control a device using a keyboard and a mouse
> because on some devices impossible to enable an additional option
> [Genymobile/scrcpy#70 (comment)](https://github.com/Genymobile/scrcpy/issues/70#issuecomment-373286323)

AFAIK this problem affects only xiaomi devices and is fixable by enabling **two** items in developer options.

**Reply by krishtoautomate** [reply]
can scrcpy-server accept hid commands if node is sending HID commands to web-sockets?

**Reply by drauggres** [reply]
It is not scrcpy-server, but Android itself handles commands.
Scrcpy client emulates physical mouse/keyboard.

**Comment by krishtoautomate** [comment]
will hid commands work on ios too?

**Reply by drauggres** [reply]
I don't know if it's possible to emulate HID for iOS (even in theory), but scrcpy uses [Android Open Accessory Protocol](https://source.android.com/docs/core/interaction/accessories/aoa2#hid-support) which is platform specific.

**Reply by krishtoautomate** [reply]
I am able to connect wired mouse and keyboard to ios device and it is working without need of installing any software

**Reply by drauggres** [reply]
AFAIK
When you connect mouse/keyboard to your device, the device acts as host (see [USB OTG](https://en.wikipedia.org/wiki/USB_On-The-Go)).
But when you connect device to a PC, device acts as client.
For android there is AOAv2 protocol which allows you to register HID. And look like there is nothing similar for iOS, I found only https://github.com/mringwal/hid-support, but it is old and requires jailbreak
